plugins/net_plugin/net_plugin.cpp

```diff
@@ -3947,103 +3950,89 @@ namespace eosio {
-                     ("cid", cid)("n", ptr->block_num())("id", id.str().substr(8,16))("m",ex.to_string()));
-         } catch( const fc::exception& ex ) {
-            exception = true;
-            fc_ilog( logger, "bad block exception connection - ${cid}: #${n} ${id}...: ${m}",
-                     ("cid", cid)("n", ptr->block_num())("id", id.str().substr(8,16))("m",ex.to_string()));
-         } catch( ... ) {
-            exception = true;
-            fc_wlog( logger, "bad block connection - ${cid}: #${n} ${id}...: unknown exception",
-                     ("cid", cid)("n", ptr->block_num())("id", id.str().substr(8,16)));
+      // proper_svnn_block_seen is for integration tests that verify low number of `unlinkable_blocks` logs.
+      // Because we now process blocks immediately into the fork database, during savanna transition the first proper
+      // savanna block will be reported as unlinkable when lib syncing. We will request that block again and by then
+      // the main thread will have finished transitioning and will be linkable. This is a bit of a hack but seems
+      // like an okay compromise for a condition, outside of testing, will rarely happen.
+      static bool proper_svnn_block_seen = false;
+
+      std::optional<block_handle> obh;
+      bool exception = false;
+      fork_db_add_t fork_db_add_result = fork_db_add_t::failure;
+      bool unlinkable = false;
+      sync_manager::closing_mode close_mode = sync_manager::closing_mode::immediately;
+      try {
+         if (cc.is_producer_node()) {
+            EOS_ASSERT(ptr->timestamp < (fc::time_point::now() + fc::seconds(7)), block_from_the_future,
+                       "received a block from the future, rejecting it: ${id}", ("id", id));
          }
-         if( exception || unlinkable) {
-            const bool first_proper_svnn_block = !proper_svnn_block_seen && ptr->is_proper_svnn_block();
-            if (unlinkable && !first_proper_svnn_block) {
-               fc_dlog(logger, "unlinkable_block ${bn} : ${id}, previous ${pn} : ${pid}",
-                       ("bn", ptr->block_num())("id", id)("pn", block_header::num_from_id(ptr->previous))("pid", ptr->previous));
-            }
-            boost::asio::post(c->strand, [c, id, blk_num=ptr->block_num(), close_mode]() {
-               peer_dlog( c, "rejected block ${bn} ${id}", ("bn", blk_num)("id", id) );
-               my_impl->sync_master->rejected_block( c, blk_num, close_mode );
-            });
-            return;
+         // This mutex is not about thread-safety; the call to accept_block is thread-safe.
+         // This mutex is to avoid extra work of validating the same block header on multiple threads at the same time.
+         // The mutex is used instead of posting to the dispatcher strand as the post to the dispatcher strand adds
+         // a thread switch which adds a delay in processing.
```

Comment:
That's going to be true for post() but quite possibly dispatch() solves that problem? Both the dispatcher strand and connection strand run on the same thread pool (same `io_context`). So I would have expected "`dispatcher_strand.dispatch()`" to immediately call the callback without a thread hop (unless the dispatcher strand is already executing something else at the moment).

Reply:
Thanks. It looks like `dispatch()` does solve the problem.

Reply:
I think in many ways what you've done with the mutex is kinda sorta what the dispatch ultimately does for you. It LGTM as is if you like it this way. There is a small possibility the dispatch will be more efficient than the mutex when the mutex is contested because a contested mutex generally implies a thread stop/start, but this stop/start may not be the case for when dispatch posts because whatever thread the strand is currently running on could pick up the post without and start/stop of a thread since that thread is already running (all this would depend on asio's impl that I'm not sure about so this could be an optimistic take).

Reply:
In my bit of testing it does seem like the `dispatch()` is better than the mutex for the nominal case. Plus it is simpler than adding a mutex.

Reply:
Am I right that this PR has two unrelated changes:
1. The switch from `post()` to `dispatch()`,  which avoids a thread hop
2. All the rest of the PR, which implements a better way to avoid validating the same block twice (although I think it could still happen, since two threads can check in `create_block_handle` that `id` is not in `fork_db`, then proceed to validate `id` before adding it to `fork_db`)

Reply:
The original dispatcher `post()` was to avoid two block header validations at the same time. This PR is an alternative, more efficient mechanism to achieve the same thing.

It is possible via the `create_block_handle` to have more than one thread validating the same block header. That is why the net_plugin serializes the calls to `create_block_handle` to avoid that. In practice, it is only called from net_plugin.

Reply:
That would work but seems rather complicated compared to current implementation.

Reply:
It would also allow to validate blocks (including signatures) in parallel, although we'd still have to add them to `fork_db` in order. Not sure if that would increase syncing performance or not though.